0020-governance-delegator-votes.md

```diff
@@ -0,0 +1,134 @@
+# ADR 0020: Governance Support Delegators Votes
+
+## Component
+
+Oasis Core
+
+## Changelog
+
+- 2022-11-02: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+With the current governance voting mechanism ([ADR-0004]), only the active
+validator set is participating in voting. This means that the validators are
+voting on behalf of all their delegators. This ADR proposes a change so that
+each delegator is able to vote with its own stake. The delegators vote acts as
+an override of the validator vote.
+
+## Decision
+
+### Casting Votes
+
+In the current implementation the submitter of a vote needs to be a part of the
+active validator committee at the time the vote is cast. This requirement is
+relaxed so that additionally anyone with a delegation to an active validator
+committee entity can vote.
+
+This change requires an efficient `staking.DelegationsFor` query to obtain a
+list of accounts the submitter is delegating to. Staking state is updated with:
+
+```go
+
+// delegationKeyReverseFmt is the key format used for reverse mapping of
+// delegations (delegator address, escrow address).
+//
+// Value is CBOR-serialized delegation.
+delegationKeyReverseFmt = keyformat.New(0x5A, &staking.Address{}, &staking.Address{})
+```
+
+`state.SetDelegation` function is updated to store both `delegationKeyFmt` and
+the reverse `delegationKeyReverseFmt`.
+`DelegationsFor` query function is updated to use the added reverse mapping.
+
+For completeness the same can be done for debonding delegations, although not
```

Comment:
@pro-wh

Reply:
lol

Reply:
Yeah  for the debonding delegations the other way could be added (making `DebondingDelegationsTo` efficient).